js/src/Ice/InputStream.js

```diff
@@ -928,98 +910,75 @@ class PatchEntry {
-        // (buffer)
-        // (communicator, encoding)
-        // (instance, encoding)
-        // (communicator, array)
-        // (instance, array)
-        // (communicator, buffer)
-        // (instance, buffer)
-        // (communicator, encoding, array)
-        // (instance, encoding, array)
-        // (communicator, encoding, buffer)
-        // (instance, encoding, buffer)
-        // (encoding, array)
-        // (encoding, buffer)
-        //
-        arr.forEach(arg => {
-            if (arg !== null && arg !== undefined) {
-                if (arg.constructor === Communicator) {
-                    args.instance = arg.instance;
-                } else if (arg.constructor === Instance) {
-                    args.instance = arg;
-                } else if (arg.constructor === EncodingVersion) {
-                    args.encoding = arg;
-                } else if (arg.constructor === Buffer) {
-                    args.buffer = arg;
-                } else if (arg.constructor === ArrayBuffer) {
-                    args.bytes = arg;
-                } else if (arg.constructor === Uint8Array) {
-                    args.bytes = arg.buffer;
+        // The first argument must be a Communicator or an Instance.
+        if (arg1 === null || arg1 === undefined) {
+            throw new InitializationException("missing argument(s) for InputStream constructor");
+        }
+
+        if (arg1.constructor === Instance) {
+            // (instance, encoding) or (instance, encoding, buffer)
+
+            this._instance = arg1;
+            if (arg2 === null || arg2 === undefined || arg2.constructor !== EncodingVersion) {
+                throw new InitializationException(
+                    "expected the encoding version as the second argument to the InputStream constructor",
+                );
+            }
+            this._encoding = arg2;
+
+            if (arg3 === undefined || arg3 === null) {
+                this._buf = new Buffer();
+            } else {
+                if (arg3.constructor === Buffer) {
+                    this._buf = arg3;
                 } else {
                     throw new InitializationException("unknown argument to InputStream constructor");
                 }
             }
-        });
-        if (args.buffer !== null && args.bytes !== null) {
-            throw new InitializationException("invalid argument to InputStream constructor");
+        } else if (arg1.constructor === Communicator) {
+            // (communicator, encoding, buffer) or (communicator, buffer), with two flavors for buffer
+            this._instance = arg1.instance;
+
+            [arg2, arg3].forEach(arg => {
+                if (arg !== null && arg !== undefined) {
+                    if (arg.constructor === EncodingVersion) {
+                        if (this._encoding !== undefined) {
+                            throw new InitializationException("duplicate encoding argument to InputStream constructor");
```

Comment:
maybe clear to say "expected the buffer third argument to the InputStream constructor" like above